src/bt.rs

```diff
@@ -74,324 +73,217 @@ pub enum BinaryTreeError<V> {
-                Ref::Other(Some(n)) => Ref::Other(if !bit { &mut n.left } else { &mut n.right }),
-                Ref::Other(None) => {
-                    return Err(BinaryTreeError::UnreachableNode(value));
+            match *node {
+                None => return Err(BinaryTreeError::UnreachableNode(value)),
+                Some(next) => {
+                    let n = &mut self.nodes[next];
+                    node = if !bit { &mut n.left } else { &mut n.right };
                 }
-            };
+            }
         }
 
-        // Checks whether the node already has a value,
-        // If so, returns an error wrapping the value that could not be inserted.
-        // otherwise, a new node is created containing the value to be inserted.
-        match node {
-            Ref::This(_) | Ref::Other(Some(_)) => {
-                return Err(BinaryTreeError::InsertNonEmptyNode(value))
-            }
-            Ref::Other(empty_node) => {
-                *empty_node = Some(Box::new(Node::new(value)));
+        if node.is_some() {
+            return Err(BinaryTreeError::InsertNonEmptyNode(value));
+        } else {
+            *node = Some(last);
+
+            if self.nodes.len() == self.nodes.capacity() {
+                self.nodes.reserve(Self::NODES_CAPACITY);
             }
+
+            self.nodes.push(Node {
+                value,
+                left: None,
+                right: None,
+            });
         }
 
         Ok(())
     }
 
     /// Gets a reference to the value located at the end of the path.
     ///
-    /// This function traverses the tree from this node (`self`) until reaching the
+    /// This function traverses the tree from the root node until reaching the
     /// node at the end of the path. It returns [None], if the node is
     /// unreachable or nonexistent. Otherwise, it returns a reference to the
     /// value stored in the node.
     pub fn get(&self, path: &Path) -> Option<&V> {
-        let mut node = self;
+        let mut node = self.root;
         for bit in path {
-            match if !bit { &node.left } else { &node.right } {
-                None => return None,
-                Some(next_node) => node = next_node,
-            };
+            let next = &self.nodes[node?];
+            node = if !bit { next.left } else { next.right };
         }
-        Some(&node.value)
+
+        Some(&self.nodes[node?].value)
     }
 
     /// Gets a mutable reference to the node located at the end of the path.
     ///
-    /// This function traverses the tree from this node (`self`) until reaching the
+    /// This function traverses the tree from the root node until reaching the
     /// node at the end of the path. It returns [None], if the node is
     /// unreachable or nonexistent. Otherwise, it returns a mutable reference
     /// to the node.
     pub fn get_node(&mut self, path: &Path) -> Option<&mut Node<V>> {
```

Comment:
Q. What is the intended use of this function? It will return a mutable reference to a `Node`, but all of `Node`'s fields are private and `Node` has no publicly-accessible functionality other than an `Encode` implementation.

Reply:
There used to be a `Node::insert()`, among others. The idea was that lookups and insertions starting from an intermediate node could skip the first part of tree traversal by retaining a reference to a node. I think this would have to be split out to a new type, like `NodeRef<'a, V>`, to work with the new arena-based approach without running afoul of borrowck.

Reply:
Now, I am using `NodeRef` that store the index of the node in the tree. This reference is returned every time a node is inserted. In addition, the `insert_at` method can take a node reference so during insertion, the traversal can start from other node different than the root.